### PR TITLE
Accept tuple.0.0 as tuple indexing (take 2)

### PR DESCRIPTION
**Pull request by petrochenkov**
If we expect something identifier-like when parsing a field name after `.`, but encounter a float token, we break that float token into parts, similarly to how we break `&&` into `&` `&`, or `<<` into `<` `<`, etc.

An alternative to https://github.com/rust-lang/rust/pull/70420.

### CODE REVIEW

File: src/test/ui/parser/float-field.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
+struct S(u8, (u8, u8));
+
+fn main() {
+    let s = S(0, (0, 0));
+
+    s.1e1; //~ ERROR no field `1e1` on type `S`
+    s.1.; //~ ERROR unexpected token: `;`
```
Comment (bjorn3) [inline src/test/ui/parser/float-field.rs:7]: Can you test `s.1. 1`?

**Comment by nikomatsakis** [comment]
Nominating for @rust-lang/lang to discuss in upcoming meeting. (This was created at our request, see discussion in #70420.)

**Comment by CAD97** [comment]
<sub>(I don't know where else to put it but it felt relevant to note somewhere. You mention the possibility of only lexing decomposed floating point literals (i.e. `1.2` would be `lit_int dot lit_int` and not `lit_float`) and gluing floating point literals together in the parser. [I'm taking this approach in an experimental formal lexical grammar](https://github.com/CAD97/rust-lexical-spec), and it works, but is not without its fun edge cases. The most relevant is `0.0b·`, where `·` is an XID_Continue character. Without floating point in the lexer, this naively lexes as `lit_int dot lit_bin_int ERROR`.)</sub>

<sub>And as a side side note: don't decompose away the `e` exponent of float literals or the `0b`/`0o`/`0x` of non-base-10 integer literals; that way lies madness and very ugly identifier introspection.</sub>

<sub>As such,</sub> As a member of wg-grammar (but not speaking for the team), I'm _somewhat_ wary of the impact this will have on an eventual syntactical specification. I don't think our current draft can split `proc_macro` tokens in the way required to implement this. **However**, I still think that the intent here is a net positive, and this token splitting seems like the least bad way to get this short term (i.e. before full decomposed tokens throughout).

**Comment by petrochenkov** [comment]
@CAD97 
>And as a side side note: don't decompose away the e exponent of float literals or the 0b/0o/0x of non-base-10 integer literals; that way lies madness and very ugly identifier introspection.

:)
I didn't thought about splitting `0b/0o/0x`, but lexically `e` seems to be just a integer token suffix like e.g. `u8` (or any other currently invalid suffix reserved for future use).
To understand what part of `0xabcdefu8` is a suffix we need to lex until we encounter something outside of the `[a-f]` range. I guess that's the ugly introspection?

**Comment by CAD97** [comment]
@petrochenkov It _might_ be easier if the entire stack is fully decomposed to treat `1e10f32` as a regular suffix. But the most annoying part for my experimental specification is that e.g. `0o0e` is rejected by the lexer as an octal float, but if it were just a literal suffix, that would be accepted. _Plus_ exponentiation is `[eE][+-]?`, so it can't just be handled by an integer suffix to a numeric base stem.

Treating `0[box]` nondecimal integers as an integer suffix is annoying because then you have to introspect the integer token to see if it's a zero, in addition to checking the integer suffix for what looks like the rest of the literal.

In general, though, if you're doing decomposed tokens, it's really not great if you have to re-introspect tokens more granularly, except for the special case of reserved words (keywords). Eating the entire non-decimal integers and `e`-based floats in the raw lexer layer seems to be a lot more convenient so far. To not derail this PR further, if you want to talk about this further, you can open a thread [over at the PR where I'm working on the experimental spec](https://github.com/CAD97/rust-lexical-spec) or [message me on zulip](https://rust-lang.zulipchat.com/#narrow/pm-with/132829-user132829). I'm open to being corrected 😄

**Comment by joshtriplett** [comment]
Dropping nomination. Based on discussions in today's lang team meeting, we're fine with this approach, and we're happy to delegate further iteration and implementation to the compiler team.

**Comment by nikomatsakis** [comment]
I would note a few things that we said in the lang team meeting but which weren't in Josh's summary, but I think they're worth noting:

* I think this approach is clever, but it's going to push a bit of a burden onto all Rust parser implementations (e.g., syn will have to reproduce this sort of logic as well).
* Modifying the lexer to use 'compound tokens' for floats (i.e., `1.0` produces a `integer` and then a `dot` and then another `integer`, with the "no whitespace" flag , or whatever scheme we wound up with for joint tokens) seems like it would've been the right approach, but it's backwards incompatible and would break existing code.
* The approach @dtolnay used initially, I think, was to modify the lexer definition so that `0.0` lexes differently when preceded by a `.`, right? Or something like that? (I didn't find a very precise definition in the PR)

I guess that upon further reflection I'm not sure that putting this burden on the *parser* is the right call. It's going to effect our ability to make a formal grammar, and it's going to have to be replicated by all other parsers? 

We could plausibly handle the backwards compatibility option by having the lexer have a flag, or by adopting the "smarter lexer" approach that I think @dtolnay originally proposed (since all proc-macros and the like would then benefit from using our lexer automatically). 

I think in the end I'm comfortable with whatever approach we take, but I do think that we should

(a) have a write-up that expands on the different approaches considered
(b) summarize the pros/cons of each approach and the impacts on procedural macro authors and parser authors
(c) lays out the reason to pick one particular approach -- that's a bit what I'm missing.

As always, I don't think this has to be a treatise -- this comment is maybe a good part of the way there? But I think it'd be useful to write out the logic.

**Comment by nikomatsakis** [comment]
Also, not to be a procedural stickler, but I am thinking that we really *ought* to be doing a FCP here, right? This is a visible change to the language and I think it merits a FCP to give folks a "heads up".

@petrochenkov do you think you could try a write-up like I described so I can fcp merge?

**Comment by petrochenkov** [comment]
>do you think you could try a write-up like I described so I can fcp merge?

Yes, probably next week.

**Comment by CAD97** [comment]
@nikomatsakis 

> [Using decomposed float tokens is] backwards incompatible

I don't think this is necessarily true.

Using decomposed tokens for floats _in `proc_macro::TokenStream` or `macro_rules!`'s `$:tt` would_ be, but using a decomposed model for `rustc_lexer` and `rustc_parse` is not, so long as they're glued back together for the above stable APIs.

(That said: `proc_macro` based parsers (e.g. `syn`) would still have to do token splitting anyway, unless the gluing was done with the "smarter lexer" (informal) context. Effectively, it's requiring the `macro_rules!`-style operator gluing for float literals. Final side note: I'm exploring decomposing the literal suffix off of the literal literal in my micro lexer, so a design following that would have to have this gluing-for-`proc_macro` anyway.)

It depends on how close you want `librustc_parse` to be to `proc_macro`, I suppose. Or if you're willing to have `librustc_parse` diverge slightly from `proc_macro` and introduce a new `meta::tokens::TokenStream` that is more fully decomposed in the future. (`proc_macro::TokenStream` -> `meta::tokens::TokenStream` would be lossless (given a span splitting API) if it continues to not include whitespace explicitly.)

_Looong_ term, I think properly (always) decomposing floats in the lexer is the better choice, but would require a breaking `proc_macro -> meta` transition to expose fully to `proc_macro` consumers. However, if (and only if) always representing floats as decomposed in the lexer token stream is infeasable, then I think adding another "smarter lexer" rule for this is probably less technical debt long term, because it's fully contained within the lexer (like the `ident dot identifier` rule; I think that most people would agree that having that split is _way_ better).

(I can just say that my micro lexer experiment would not like it if it has to handle the "smarter lexer" option in its current design; the existing "smarter lexer" behavior for `integer dot ident` is the most complicated part of the definition as currently written, and `dot ident $(dot ident)+`, as currently specified, would take an infinite amount of rules. (It's fixable but would take a complete restructuring of the current formalism.))

TL;DR we can stability without stagnation our lexer if we just always shim to/from the stable `proc_macro` API. (Note: said shim will still be required if/when the "IDE syntax tree unification" happens (cc @matklad).)

**Comment by petrochenkov** [comment]
>TL;DR we can stability without stagnation our lexer if we just always shim to/from the stable `proc_macro` API.

And the shim already exists due to translation of doc comments and maybe some other stuff, it just works only in the opposite right now - "one `rustc` token -> multiple `proc_macro` token", rather than "multiple `rustc` tokens <-> one `proc_macro` token".

**Comment by nikomatsakis** [comment]
@CAD97 Indeed. I believe the approach you're suggesting is what I was hinting at when I wrote this:

> We could plausibly handle the backwards compatibility option by having the lexer have a flag, or by adopting the "smarter lexer" approach that I think @dtolnay originally proposed (since all proc-macros and the like would then benefit from using our lexer automatically).

In particular, we can clearly preserve the lexing behavior for procedural macros by inserting some shims to convert from the new tokens to the old ones -- and I think that's fine. But it's also true that procedural macros would probably **prefer** the tokens that rustc is seeing, since it would enable e.g. `syn` to parse Rust more easily (i.e., the same reason that our parser wants different tokens). 

I guess this is what you meant when you talked about the "meta" shift -- i.e., we might use that shift as a kind of "edition" change for procedural macros?

**Comment by petrochenkov** [comment]
## Stabilization report

This PR implements parsing `tuple.0.0` as two field accesses instead of erroneous `tuple`, `.` and floating point `0.0`.
The feature is implemented as insta-stable.

To be more specific, if we expect something identifier-like when parsing a field name after `.`, but encounter a float token, we break that float token into parts, similarly to how we break `&&` into `&` `&`, or `<<` into `<` `<`, etc.

## Larger picture

There are two ways to achieve `tuple.0.0` being parsed as two tuple field accesses.
- One way is to keep the current "compound" float tokens and introduce a special case into the lexer so it treats `0.0` differently depending on its context (as one float token, or as three integer and punctuation tokens).
- Another way is to break the float tokens into more basic components (possibly suffixed integer tokens and punctuation) and defer the task of discerning between floats and tuple field accesses to the parser.

Pros and cons of both approaches:
- Compound float tokens
	- (Negative) From the general design point of view, if we have basic components which we want to break into parts sometimes (like `0.0` tokens), perhaps they are not basic enough and we need to rethink our set of basic components. Especially given that breaking floats doesn't require introducing any *new* basic components, since they can be broken into already existing integers and punctuation.
	- (Positive) Freezing the tokens produced by rustc lexer (which provides no stability guarantees) allows to avoid any compatibility issues with proc macro APIs and `macro_rules` matchers which do provide stability guarantees.
	- (Positive) There's some precedent for treating float tokens specially in the lexer, `0.` is lexed as two tokens rather than as a float if it's followed by `.` or an identifier (to support ranges `0..1` and integer methods `0.foo()`).
- Breaking float tokens into integers and punctuation
	- (Positive) The general design point of view, see above. Compound operators (like `<<`) are currently represented as single tokens in rustc lexer and there are already plans to break them up with the same motivation.
	- (Positive) Can remove the existing floating-point-related special cases in rustc lexer (`0..1` and `0.foo()`).
	- (Negative) Proc macro API and `tt` matcher have stability guarantees, we will need shims at proc and decl macro boundaries to represent multiple internal tokens as one (`0.0` or `<<`) when matching with `tt` and when producing proc macro tokens. (Bumping the major version of proc macro API and the ways to make migration from pmAPI1 to pmAPI2 painless are out of scope right now.)
	- (Negative?) Lexer does less work, so parsers have to do more work. There are multiple parsers, so they all will need to provide tools for combining multi-character operators (`<` `<` -> `<<`) and floating point literals (`0` `.` `0` -> `0.0`, assuming proc macro API migrates to the new model) from basic tokens, or breaking float tokens into parts (`0.0` -> `0` `.` `0`, assuming proc macro API is frozen). `proc_macro` crate itself doesn't provide such facilities, so they need to be implemented by crates like `syn` that provide conveniences on top of the barebones `proc_macro`.

This PR chooses the second strategy (breaking float tokens), but emulates it in the parser, similarly to how we emulate future single-character punctuation by breaking `<<` into `<` and `<` in the parser.
Later when we have some necessary infrastructural work done we migrate both `<<` and `0.0` to more basic tokens properly in the lexer.

**Comment by petrochenkov** [comment]
@rust-lang/lang 
Could someone start an FCP?

**Comment by nikomatsakis** [comment]
@petrochenkov 

Before I do, a point of clarification. Under the heading "Compound float tokens", you wrote

> (Positive) There's some precedent for treating float tokens specially in the lexer, `0.` is lexed as two tokens rather than as a float if it's followed by `.` or an identifier (to support ranges `0..1` and integer methods `0.foo()`).

But this doesn't sound like using a compound float token, unless I'm misunderstanding. In particular, you wrote that `0.` is *lexed* in different ways, whereas to be precedent, I would expect `0.` to always be lexed as a floating point number and then to have the parser break it up into subtokens.

**Comment by nikomatsakis** [comment]
@petrochenkov 

One other point:

> (Negative?) Lexer does less work, so parsers have to do more work. There are multiple parsers, so they all will need to provide tools for combining multi-character operators

This is an interesting point, though I think I goes both ways. That is, the approach we're currently taking requires all parsers to understand that `0.0` might need to be "broken up" into subtokens, whereas the other approach requires all parsers to be able to combine `0`, `.`, and `0` into a floating point number (modulo whitespace). To me, this comes out *slightly* in favor of a simpler lexer, since combining simple subtokens into more complex things is kind of "what parsers do", whereas taking a single token and breaking it apart is quite an unusual concept (it's not part of any parser formalism I'm aware of).

One other question: For float tokens like `0.0_f32` and so forth, we obviously want to accept that as field accesses... are suffixes like `_f32` and so forth part of the token? That seems worth noting somewhere, though it's not part of the "core tradeoff".

**Comment by nikomatsakis** [comment]
> This PR chooses the second strategy (breaking float tokens), but emulates it in the parser

This seems a bit confusing to me -- it seems like the PR chooses the **first strategy**, in that the lexer still produces compound tokens.

**Comment by CAD97** [comment]
The first strategy is the one in which `foo.0.0` lexes as `ident dot digits dot digits` but also lexes `foo. 0.0` as `ident dot whitespace float` (i.e. "smarter lexer").

The second strategy is where `foo.0.0` lexes as `ident dot digits dot digits` and `foo. 0.0` lexes as `ident dot whitespace digits dot digits`.

This PR chooses the second approach, but the implementation is sort of halfway in-between, as the lexer still produces `ident dot float`. It's just that we've now taught the parser to tear the `float` lexical token into its component tokens, the same way that e.g. `>>` is handled.

> `0.0_f32`

To be clear, the "micro" lexical tokens for this would be

- `digits` `0`
- `dot` `.`
- `digits` `0_`
- `ident` `f32`

if literal suffixes are also decomposed. (Because (IIUC) there are no cases where a suffixed literal makes sense as an unsuffixed literal followed by an identifier, you could also choose to lex the suffix as part of the identifier.) Currently, it always lexes as a single token, `float 0.0_f32 (suffix f32)`.

> we obviously want to accept that as field accesses

Really? #60210 is a compatibility issue specifically about phasing it out as it was incorrectly accepted.

**Comment by CAD97** [comment]
Given #60210, here's two new tests to test:

```rust
fn main() {
    let foo = ((0,),);
    dbg!(foo.0.0usize);
}
```

(Currently unaccepted. Should mention #60210 in the error (or maybe warning?) after this patch. See the error for unacceptable suffixes, e.g. `foo.0u`.)

```rust
fn main() {
    let foo = ((0,),);
    dbg!(foo.0usize.0usize);
}
```

(Currently accepted, warns of #60210 future-compat. Certainly unaffected by this patch, but tangentially relevant.)

**Comment by nikomatsakis** [comment]
> Really? #60210 is a compatibility issue specifically about phasing it out as it was incorrectly accepted.

Thanks -- I meant to write *we obviously do NOT want*... will edit.

**Comment by petrochenkov** [comment]
@nikomatsakis
>But this doesn't sound like using a compound float token

I've changed "Compound float tokens" to "Compound float tokens in general, broken in the lexer in few special cases".
The example you mentioned is more about the "broken in the lexer in few special cases" part.
The second approach doesn't have such special cases in the lexer.

>This seems a bit confusing to me -- it seems like the PR chooses the first strategy, in that the lexer still produces compound tokens.

Same as above.

**Comment by petrochenkov** [comment]
r? @nikomatsakis 
Did me and CAD97 answered your questions?

**Comment by nikomatsakis** [comment]
@petrochenkov yes I found the newer version easier to follow, thanks

**Comment by nikomatsakis** [comment]
I noticed @petrochenkov you tagged this with T-compiler and T-lang -- I was going to kick-off an FCP, but I think only for T-lang, did you feel it should be a T-compiler FCP?

**Comment by petrochenkov** [comment]
@nikomatsakis 
Well, I added the T-compiler tag long before the FCP, because the change is related to the plans for reforming token treatment in the compiler.
I don't have an opinion on whether this should be a T-lang FCP or both, T-lang would probably be enough.

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

I am proposed to [accept this change](https://github.com/rust-lang/rust/pull/71322#issuecomment-629835568) that enables us to parse as `tuple.0.0`. As the stabilization report states, the current change does not affect the tokens produced by the lexer or provided to procedural macros, but instead modifies the parser to "split" tokens like `0.0` into their components on demand.

**Comment by nikomatsakis** [comment]
Tagging for release notes as this is a noteworthy change.

**Comment by dtolnay** [comment]
I would love a PR if anyone is interested in helping implement this in Syn. https://github.com/dtolnay/syn/issues/844

**Comment by nikomatsakis** [comment]
ping @cramertj @scottmcm and @withoutboats [on this FCP](https://github.com/rust-lang/rust/pull/71322#issuecomment-637632991)

File: src/test/ui/parser/float-field.rs (hunk 1 above)
Comment (petrochenkov) [reply src/test/ui/parser/float-field.rs:7]: Done.

**Review by dtolnay — COMMENTED** [review]
I implemented the corresponding parser change in syn 1.0.33.

I am in favor of stabilizing this as implemented here but I would flag one unfortunate behavior not mentioned in the stabilization report. According to @petrochenkov in https://github.com/rust-lang/rust/pull/70420#issuecomment-604643571:

> With the scheme from my comment above `$expr.$float` won't work because `$float` is not a float token, it's an "interpolated" token, it basically have invisible parentheses around it.

`$expr.$float` not working is the desired behavior (PR should add a test), but while it may be the case for rustc's parser, syn and proc macros in general have no way to make the same distinction. In the proc macro below, `$expr.$float` appears indistinguishable from `tuple.0.0` in the input token stream and the proc macro can only accept both or neither.

```rust
// lib.rs

// [dependencies]
// syn = { version = "1.0", features = ["full", "extra-traits"] }

use proc_macro::TokenStream;
use syn::{parse_macro_input, Expr};

#[proc_macro]
pub fn repro(input: TokenStream) -> TokenStream {
    let expr = parse_macro_input!(input as Expr);
    println!("{:#?}", expr);
    TokenStream::new()
}
```

```rust
// main.rs

macro_rules! m {
    ($expr:expr, $float:tt) => {
        repro!($expr.$float);
    };
}

m!(tuple, 0.0);

fn main() {}
```

**Comment by petrochenkov** [comment]
@dtolnay 
If `$float` is a `tt` like in the example above (as opposed to `expr` or `literal`), then `$expr.$float` will work in `rustc` as well.

`macro_rules` substitute `tt`s very early and `tt`s are not wrapped into delimited groups by design, so neither rustc parser nor proc macro parsers can see them.
If `$float` is any other matcher, like `expr`, then rustc parser will recognize it as `NtExpr` and proc macro parsers can recognize it because it's wrapped into a `Group`.

**Comment by nikomatsakis** [comment]
Personally I would expect `$expr.$float` to work, given the proposed design. I can see why one might not *want* it to work, but it matches my mental model that the "lexer produces a float, parser consumes a partial token and pushes remaining tokens back into the stream".

**Comment by jplatte** [comment]
Should this have the relnotes tag and a milestone?
